chapter_02B_abstractions.asciidoc

```diff
@@ -631,71 +581,65 @@ def ej_synchronise_dirs(reader, filesystem, source_root, dest_root):
 [source,python]
 [role="skip"]
 ----
+class FakeFileSystem(list): #<1>
+
+    def copy(self, src, dest): #<2>
+        self.append(('COPY', src, dest))
+
+    def move(self, src, dest):
+        self.append(('MOVE', src, dest))
+
+    def delete(self, dest):
+        self.append(('DELETE', src, dest))
+
+
 def test_when_a_file_exists_in_the_source_but_not_the_destination():
     source = {"sha1": "my-file" }
     dest = {}
-    actions = []
+    filesystem = FakeFileSystem()
 
     reader = {"/source": source, "/dest": dest}
-    synchronise_dirs(reader.pop, actions.append, "/source", "/dest")
+    synchronise_dirs(reader.pop, filesystem, "/source", "/dest")
 
-    assert actions == [("COPY", "/source/my-file", "/dest/my-file")]
+    assert filesystem == [("COPY", "/source/my-file", "/dest/my-file")]
 
 
 def test_when_a_file_has_been_renamed_in_the_source():
     source = {"sha1": "renamed-file" }
     dest = {"sha1": "original-file" }
-    actions = []
+    filesystem = FakeFileSystem()
 
     reader = {"/source": source, "/dest": dest}
-    synchronise_dirs(reader.pop, actions.append, "/source", "/dest")
+    synchronise_dirs(reader.pop, filesystem, "/source", "/dest")
 
-    assert actions == [("MOVE", "/dest/original-file", "/dest/renamed-file")]
+    assert filesystem == [("MOVE", "/dest/original-file", "/dest/renamed-file")]
 ----
 ====
 
-////
-TODO (ej)
-Modified this a bit to also introduce the filesystem abstraction.
-
-class FakeFilesystem(object):
-    def __init__(self, actions):
-        self._actions = actions or []
-
-    def copy(self, src, dest):
-        self._actions.append(('COPY', src, dest))
-
-    def move(self, src, dest):
-        self._actions.append(('MOVE', src, dest))
-
-    def delete(self, dest):
-        self._actions.append(('DELETE', src, dest))
-
-
-then
-
-
-     actions = [] # Using the shunt pattern here.
-
-////
+<1> Bob _loves_ using lists to build simple test doubles, even though his
+co-workers get mad. It means we can write tests like
+`assert 'foo' not in database`
 
+<2> Each method in our `FakeFileSystem` just appends something to the list so we
+can inspect it later. This is an example of a Spy Object.
 
 
 The advantage of this approach is that your tests act on the exact same function
-that's used by your production code.  The disadvantage is that DI usually demands
-a bit more work on the part of the reader to understand what's going on.
+that's used by your production code.  The disadvantage is that we have to make
+our stateful components explicit and we have to pass them around. DHH famously
+described this as "test damage".
```

Comment:
I'm sure that some readers would be asking at this point, "Why bother with the fake when you could just use a unittest.mock?"  This might be a more natural spot to put the blurb about London vs Detroit style testing.

Reply:
i was also thinking today that this chapter might be a nice place for a little sidebar saying "did you notice we didn't use any mocks?  let's talk about that."